### PR TITLE
[JUJU-1513] migrate assess_upgrade_series.py to bash

### PR DESCRIPTION
**Pull request by barrettj12**
The `nw-upgradeseries-amd64-lxd` is constantly failing. Hence, now's a good time to migrate it into the bash test suites.

## QA steps

```console
cd tests
./main.sh -v upgrade_series
```

## TODO
- [x] rebase to target 2.9 instead, `nw-upgradeseries-amd64-lxd` is failing there as well
- [x] `dummy-sink` gets stuck in maintenance status after upgrade - fix this
- [x] test on 2.9
- [x] test on 3.0
- [ ] make some corresponding changes to `juju-qa-jenkins`

### CODE REVIEW

**Comment by barrettj12** [comment]
I can't get the `dummy-sink` charm to behave properly after a reboot:
```bash
juju add-model m
juju deploy ./testcharms/charms/dummy-sink
juju deploy ./testcharms/charms/dummy-source
juju relate dummy-sink dummy-source
juju config dummy-source token=foo
# wait for charms to start
juju status
# dummy-sink: active "Token is foo"

juju ssh 0 'sudo shutdown -r now'
# wait for machine to restart
juju status
# dummy-sink: blocked "Waiting for token"
```

File: tests/suites/upgrade/series.sh [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# Deploy the dummy-sink charm on an earlier series, then check that we can
+# upgrade the series, and that the charm still behaves correctly.
```
Comment (benhoyt) [inline tests/suites/upgrade/series.sh:2]: Thanks for the comments!

**Review by benhoyt — APPROVED** [review]
One small question, but this looks good to me, nice work! I was going to comment that you hadn't yet deleted the Python, but you deleted it as I was writing the comment. :-)

I ran this locally on 2.9 and it worked nicely (didn't test on 3.0 though).

Would be so much better if our Jenkins configuration was in the same repo so that could be part of the same commit ... oh well.

File: tests/suites/upgrade/series.sh [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+	# Setup
+	juju add-model test-upgrade-series
+	juju deploy ./testcharms/charms/dummy-sink --series $start_series
+	juju deploy ./testcharms/charms/dummy-source --series $end_series
+	juju relate dummy-sink dummy-source
+	juju config dummy-source token=Canonical
+
+	# Check pre-conditions
+	wait_for "Canonical" "$(workload_status "dummy-sink" 0).message"
```
[the remark this answers is not in the document]
Comment (barrettj12) [reply tests/suites/upgrade/series.sh:22]: Hah, good point. Yeah I assume it works cause of the parentheses, maybe I will change to single quotes to be safe

File: tests/suites/upgrade_series/series.sh [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	if [[ $juju_major_version == "3" ]]; then
+		start_series="focal"
+		end_series="jammy"
+	else
+		start_series="bionic"
+		end_series="focal"
+	fi
+
+	# Setup
+  ensure "test-upgrade-series-relation" "${TEST_DIR}/test-upgrade-series-relation.log"
```
Comment (wallyworld) [inline tests/suites/upgrade_series/series.sh:15]: whitespace?

File: tests/suites/upgrade_series/series.sh [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	reboot_machine 0
+	echo "Upgrading machine..."
+	echo "See ${TEST_DIR}/do-release-upgrade.log for progress."
+	juju ssh 0 'sudo do-release-upgrade -f DistUpgradeViewNonInteractive' &>"${TEST_DIR}/do-release-upgrade.log" || true
+	reboot_machine 0
+	juju upgrade-series 0 complete
+
+	# Check post-conditions
+	wait_for "Canonical" "$(workload_status 'dummy-sink' 0).message"
+	assert_machine_series 0 $end_series
```
Comment (wallyworld) [inline tests/suites/upgrade_series/series.sh:42]: we also need to assert that the upgrade series hook has run on both charms

File: tests/suites/upgrade_series/series.sh [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	juju deploy ./testcharms/charms/dummy-sink --series $start_series
+	juju deploy ./testcharms/charms/dummy-source --series $end_series
+	juju relate dummy-sink dummy-source
+	juju config dummy-source token=Canonical
+
+	# Check pre-conditions
+	wait_for "Canonical" "$(workload_status 'dummy-sink' 0).message"
+	assert_machine_series 0 $start_series
+
+	# For Juju 3.0 we want to upgrade to jammy, but since this is not LTS yet,
```
Comment (wallyworld) [inline tests/suites/upgrade_series/series.sh:25]: ????

**Review by wallyworld — APPROVED** [review]

File: tests/suites/upgrade_series/series.sh (hunk 3 above)
Comment (barrettj12) [reply tests/suites/upgrade_series/series.sh:15]: f*king goland

**Pull request merged**